expression/constant_propagation.go

```diff
@@ -287,37 +305,276 @@ func (s *propagateConstantSolver) solve(conditions []Expression) []Expression {
 	}
 	s.propagateConstantEQ()
 	s.propagateColumnEQ()
-	for i, cond := range s.conditions {
-		if dnf, ok := cond.(*ScalarFunction); ok && dnf.FuncName.L == ast.LogicOr {
-			dnfItems := SplitDNFItems(cond)
-			for j, item := range dnfItems {
-				dnfItems[j] = ComposeCNFCondition(s.ctx, PropagateConstant(s.ctx, []Expression{item})...)
+	s.conditions = propagateConstantDNF(s.ctx, s.conditions)
+	return s.conditions
+}
+
+// PropagateConstant propagate constant values of deterministic predicates in a condition.
+func PropagateConstant(ctx sessionctx.Context, conditions []Expression) []Expression {
+	solver := &propConstSolver{}
+	solver.colMapper = make(map[string]int)
+	solver.ctx = ctx
+	return solver.solve(conditions)
+}
+
+type propOuterJoinConstSolver struct {
+	basePropConstSolver
+	jConds      []Expression
+	fConds      []Expression
+	outerSchema *Schema
+	innerSchema *Schema
+}
+
+func (s *propOuterJoinConstSolver) setConds2ConstFalse(jConds, fConds bool) {
+	if jConds {
+		s.jConds = []Expression{&Constant{
+			Value:   types.NewDatum(false),
+			RetType: types.NewFieldType(mysql.TypeTiny),
+		}}
+	}
+	if fConds {
+		s.fConds = []Expression{&Constant{
+			Value:   types.NewDatum(false),
+			RetType: types.NewFieldType(mysql.TypeTiny),
+		}}
+	}
+}
+
+// pickNewEQCondsFunc picks constant equal expression from specified conditions.
+func (s *propOuterJoinConstSolver) pickNewEQCondsFunc(retMapper map[int]*Constant, visited []bool, fConds bool) map[int]*Constant {
+	var conds []Expression
+	var condsOffset int
+	if fConds {
+		conds = s.fConds
+	} else {
+		conds = s.jConds
+		condsOffset = len(s.fConds)
+	}
+	for i, cond := range conds {
+		if visited[i+condsOffset] {
+			continue
+		}
+		col, con := validEqualCond(cond)
+		// Then we check if this CNF item is a false constant. If so, we will set the whole condition to false.
+		var ok bool
+		if col == nil {
+			if con, ok = cond.(*Constant); ok {
+				value, err := EvalBool(s.ctx, []Expression{con}, chunk.Row{})
+				terror.Log(errors.Trace(err))
+				if !value {
+					if fConds {
```

Comment:
Seems like it can be simplified to `s.setConds2ConstFalse(true, fConds)`, and the first parameter is always true.